core/state_processor.go

```diff
@@ -1345,9 +1430,20 @@ func ValidateQiTxInputs(tx *types.Transaction, chain ChainContext, db ethdb.Read
 	if tx.ChainId().Cmp(signer.ChainID()) != 0 {
 		return nil, fmt.Errorf("tx %032x has wrong chain ID", tx.Hash())
 	}
-	if len(tx.Data()) != 0 && len(tx.Data()) != common.AddressLength {
-		return nil, fmt.Errorf("tx %v emits UTXO with invalid data length %d", tx.Hash().Hex(), len(tx.Data()))
+	if len(tx.Data()) != 0 && (len(tx.Data()) != params.MaxQiTxDataLength && len(tx.Data()) != common.AddressLength) {
+		return nil, fmt.Errorf("tx %v emits UTXO with data %d not equal to either address length or MaxQiTxDataLength %d", tx.Hash().Hex(), len(tx.Data()), params.MaxQiTxDataLength)
+	}
+
+	// Wrap Qi Transaction
+	if len(tx.Data()) == common.AddressLength && !common.BytesToAddress(tx.Data()[:], currentHeader.Location()).IsInQuaiLedgerScope() {
+		return nil, fmt.Errorf("tx %v emits UTXO with contract that is not in quai ledger scope", tx.Hash().Hex())
 	}
+
+	// Qi To Quai Conversion
+	if len(tx.Data()) == params.MaxQiTxDataLength && !common.BytesToAddress(tx.Data()[2:22], currentHeader.Location()).IsInQiLedgerScope() {
```

Comment:
I think this is a bug

Reply:
why is this a bug? we are checking if the refund address is in Qi ledger scope